File: ios/Runner/Tajweed/CanonicalLexicalAuthority.swift
import Foundation

/// ADR-010 canonical lexicon is the **production** lexical authority on all
/// builds (Debug / Profile / Release). Legacy display-string alignment is no
/// longer selectable.
enum CanonicalLexicalAuthority {
  static let defaultsKey = "canonicalLexicalProductionEnabled"

  /// Always `true` — Canonical is the production lexical scorer.
  static var productionEnabled: Bool {
    get { true }
    set {
      // Ignored — Canonical is always on. Kept so older MethodChannel clients
      // and tests that assign the flag do not crash.
      NSLog(
        "[TajweedCanonical] set productionEnabled=%@ ignored (Canonical is always on)",
        newValue ? "YES" : "NO"
      )
    }
  }

  static func clearOverride() {
    UserDefaults.standard.removeObject(forKey: defaultsKey)
    UserDefaults.standard.synchronize()
  }

  /// Flutter MethodChannel may deliver Bool as NSNumber — accept both.
  static func parseEnabledArgument(_ raw: Any?) -> Bool? {
    if let b = raw as? Bool { return b }
    if let n = raw as? NSNumber { return n.boolValue }
    if let i = raw as? Int { return i != 0 }
    return nil
  }
}

File: android/app/src/main/kotlin/com/app/deenly/deenly/tajweed/CanonicalLexicalAuthority.kt
package com.rnr.deenfocus.tajweed

import android.content.Context
import android.util.Log

/**
 * ADR-010 canonical lexicon is the **production** lexical authority on all
 * builds. Legacy display-string alignment is no longer selectable.
 */
object CanonicalLexicalAuthority {
    const val PREFS_NAME = "tajweed_canonical"
    const val KEY = "canonicalLexicalProductionEnabled"

    @Volatile
    private var prefs: android.content.SharedPreferences? = null

    @Synchronized
    fun init(context: Context) {
        if (prefs == null) {
            prefs = context.applicationContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        }
    }

    /** Always `true` — Canonical is the production lexical scorer. */
    var productionEnabled: Boolean
        get() = true
        set(value) {
            // Ignored — Canonical is always on. Kept so older MethodChannel
            // clients that assign the flag do not crash.
            Log.i(
                "TajweedCanonical",
                "set productionEnabled=$value ignored (Canonical is always on)",
            )
        }

    fun clearOverride() {
        prefs?.edit()?.remove(KEY)?.commit()
    }

    fun parseEnabledArgument(raw: Any?): Boolean? {
        return when (raw) {
            is Boolean -> raw
            is Number -> raw.toInt() != 0
            else -> null
        }
    }
}
